frontend/src/pages/create_campaign/Roles/index.js

```diff
@@ -0,0 +1,297 @@
+  CreateRoleName,
+  AddQuestionButton,
+  SelectFromExistingButton,
+  QuestionTitle,
+  QuestionContent,
+} from "./rolesTab.styled";
+
+const RolesTab = (props) => {
+  const {
+    value,
+    index,
+    roles,
+    setRoles,
+    questions,
+    setQuestions,
+    roleSelected,
+    setRoleSelected,
+  } = props;
+  const [newRoleName, setNewRoleName] = useState("");
+  const [newRoleQty, setNewRoleQty] = useState(1);
+  const [anchorEl, setAnchorEl] = useState(null);
+  const open = Boolean(anchorEl);
+
+  const handleSelectFromExistingClick = (event) => {
+    setAnchorEl(event.currentTarget);
+  };
+  const handleCloseSelectFromExisting = () => {
+    setAnchorEl(null);
+  };
+
+  const selectFromExisting = (e) => {
+    // add currently selected role to set of roles related to an existing question
+    const updatedQuestions = questions.map((q) =>
+      q.id === e.currentTarget.value
+        ? { id: q.id, text: q.text, roles: q.roles.add(roleSelected) }
+        : q
+    );
+    setQuestions(updatedQuestions);
+    handleCloseSelectFromExisting();
+  };
+
+  // filter for questions pertaining to currently selected role
+  const filteredQuestions = questions.filter(
+    (q) => !q.roles.has(roleSelected) && q.text !== ""
+  );
+
+  const onRoleDelete = (e) => {
+    e.stopPropagation();
+    const newRoles = roles.filter((r) => r.id !== e.currentTarget.value);
+    if (e.currentTarget.value === roleSelected) {
+      if (newRoles.length !== 0) {
+        setRoleSelected(newRoles[0].id);
+      } else {
+        setRoleSelected(-1);
+      }
+    }
+    setRoles(newRoles);
+  };
+
+  const onQuestionDelete = (e) => {
+    // Remove question from role
+    questions.forEach(
+      (q) => q.id === e.currentTarget.value && q.roles.delete(roleSelected)
+    );
+    // If question relates to no roles, remove from questions
+    setQuestions(questions.filter((q) => q.roles.size > 0));
+  };
+
+  const handleQuestionInput = (e, id) => {
+    const newQuestions = questions.map((q) => {
+      if (q.id === id) {
+        return { id: q.id, text: e.currentTarget.value, roles: q.roles };
+      }
+      return q;
+    });
+    setQuestions(newQuestions);
+  };
+
+  const addQuestion = () => {
+    if (roleSelected === "-1") {
+      alert(
+        "Please create and select one or more roles before adding questions!"
+      );
+    } else {
+      const newID = uuidv4();
+      setQuestions([
+        ...questions,
+        { id: newID, text: "", roles: new Set([roleSelected]) },
+      ]);
+    }
+  };
+
+  const addRole = () => {
+    const newID = uuidv4();
+    if (newRoleName) {
+      setRoles([
+        ...roles,
+        {
+          id: newID,
+          title: newRoleName,
+          quantity: newRoleQty,
+        },
+      ]);
+      setNewRoleName("");
+      setNewRoleQty(1);
+      if (roleSelected === -1) {
+        setRoleSelected(newID);
+      }
+    } else {
+      alert("Role name is required!");
+    }
+  };
+
+  return (
```

Comment:
thers a LOT of components here - could you break it down into more readable parts? like you have the `roles.map` line and perhaps chuck that entire return value into another component

Reply:
Yep can definitely do that.

On that note, what is the general idea with extracting out into separate components? Is it mostly for things that will be reused, or just generally to keep each file simpler/clearer? Also if it isn't something that is likely to be reused, should it still go in the components folder, or just in the folder where it will be used?